Feature/filter editing

thespacedoctor opened the pull request:
Please remember to follow the Lasair Pull Request checklist when issuing & reviewing this pull request:

https://lsst-uk.atlassian.net/wiki/spaces/LUSC/pages/2612264961/How+to+do+Pull+Request#STEP-3%3A-Issue-a-Pull-Request-on-Github

# CHANGES

- You can now edit filters 

![image](https://user-images.githubusercontent.com/3385557/222429896-517c6f4c-5140-4433-a952-578487d6e669.png)

- Column definitions added to autocomplete modals

![Google Chrome - 127 0 0 1-8000-filters-create- 2023-03-02 at 11 59 23 AM-2](https://user-images.githubusercontent.com/3385557/222429978-d0654458-ad1b-4cf5-b163-9f59945da15b.gif)

- Text areas for writing queries now wrap neatly (no more having to force a new line)

![image](https://user-images.githubusercontent.com/3385557/222430100-cbff5f46-7142-4ede-81e5-466b1b5f0d03.png)

- Fixed bug when requesting table.* (e.g. objects.*)
- Autocomplete should now be more robust (wasn't working as expected 100% of the time)
- Kafka topic added to filter header

![image](https://user-images.githubusercontent.com/3385557/222430305-29da65de-cfa6-441d-928d-f2da186dc88a.png)

- Added a header to show user they are editing a filter

![image](https://user-images.githubusercontent.com/3385557/222430346-80ab842e-6b16-489c-af38-de276444f483.png)

thespacedoctor commented:
Changes now live on dev server

RoyWilliams commented:
Great job Dave, its very vey close now!

There was an error with the editing, when I saved the filter. It said "sqlquery_real" is undefined, so I just added line 329 in .../apps/filter_query/views.py to compute it. Now the edit/run/save sequence works.

I tried looking at filters with the admin module, but it is difficult because everything is on one line. I changed ../apps/filter_query/admin.py so that the chunks of text are shown as textarea instead of single line. This is what we had on the old system.

I pushed both of these into this branch.

(1) Duplication and deletion works for somebody elses filter, but I cannot duplicate a filter owned by myself.

(2) Next, logged in as 'su', I see the following filters, so a couple of problems ....

 FrontPage:      click and it shows correct results
 fink001:        click and it shows correct results	
 Lightweight:    module 'settings' has no attribute 'DEBUG'
 Sample Email:   module 'settings' has no attribute 'DEBUG'
 Sample Fast: 	 ...apps/db_schema/utils.py, line 86, in get_schema_for_query_selected
 hello123:       click and it shows correct results
 BLLac Outburst: click and it shows correct results

thespacedoctor commented:
## CHANGES

- You can now duplicate watchmaps/watchlists/filters if they belong to you (bug fix)

```bash
Lightweight: module 'settings' has no attribute 'DEBUG'
Sample Email: module 'settings' has no attribute 'DEBUG'
```

- The issue here was when the returned result count was over 5000. The code needed to look in the settings file to see if DEBUG was switched on or off to point to the correct readthedocs URL in the warning message. But I was looking in the wrong setting file. 

![image](https://user-images.githubusercontent.com/3385557/223086141-ffb93600-3396-49a3-b3d5-51a7bf93d224.png)

```bash
Sample Fast: ...apps/db_schema/utils.py, line 86, in get_schema_for_query_selected
```

- Fixed ^.

- I have now integrated `crossmatch_tns` into the schema browser and filter builder. 

![image](https://user-images.githubusercontent.com/3385557/223086253-a8b6d415-e455-4386-836f-6e4f985eced3.png)

- I added the correct `TNS_WATCHLIST_ID` to the settings file on the dev server. Any queries involving the `crossmatch_tns` were painfully slow so I added a few targeted indexes on the appropriate MySQL tables. Queries now run in < 1sec. I have added the index creation commands to the drop-day checklist:

```sql
ALTER TABLE `ztf`.`watchlist_hits` 
ADD INDEX `idx_name` (`name` ASC) ;
ALTER TABLE `ztf`.`watchlist_hits` 
ADD INDEX `idx_objectid_wl_id` (`objectId` ASC, `wl_id` ASC) ;
ALTER TABLE `ztf`.`crossmatch_tns` 
ADD INDEX `idx_name` (`tns_name` ASC) ;
```

@RoyWilliams want to take the new crossmatch_tns table for a spin in the filter builder?
